app/models/concerns/node_properties.rb

```diff
@@ -72,4 +73,127 @@ def raw_properties=(value)
+  # by matching the port and protocol columns.
+  #
+  # value - The new entry (Hash) we want to add to :services
+  #         (and may be :services_extra) tables
+  #
+  # Returns the updated services table
+  def add_service(value)
+    # extract extra info from value
+    value_extra = value.slice(:port, :protocol, :extra)
+    value.slice!(:name, :port, :product, :protocol, :reason, :state, :version)
+
+    merge_service(value)
+
+    unless value_extra[:extra].nil?
+      value_extra[:extra].select! { |extra| !extra[:output].blank? }
+      if value_extra[:extra].any?
+        merge_service_extra(value_extra)
+      end
+    end
+
+    self.properties[:services]
+  end
+
+  # Private: Merges a Hash into an Array of Hashes by :port and :protocol keys.
+  # If a row exists in the table with the new row port and protocol values, the
+  # rest of the columns that have the same name and different values
+  # are concatenated (as strings)
+  #
+  # table   - the existing table where we want to merge a new entry.
+  #         May be nil, a Hash or an Array of hashes
+  # new_row - the new entry (a Hash) we want to put in the existing table.
+  #
+  # Returns table with new_row merged
+  def merge_service(new_row)
+    table = self.properties[:services]
+
+    # work always with an array of hashes
+    table = [table] if table.is_a?(Hash)
+    table = [] if table.nil?
+
+    # find new_row in table
+    position = table.find_index do |row|
+      row.values_at(:port, :protocol) == new_row.values_at(:port, :protocol)
+    end
+
+    if position.nil?
+      # if the row was not in the table, add it
+      table << new_row
+    else
+      # if the row is in the table, merge it
+      target = table[position]
+      target.merge!(new_row) do |_key, oldvalue, newvalue|
+        if oldvalue == newvalue
+          oldvalue
+        else
+          [oldvalue.split(' | '), newvalue].flatten.uniq.join(' | ')
+        end
+      end
+
+      table[position] = target
+    end
+
+    self.properties[:services] = table
+  end
+
+  # Private: Merges a Hash into an Array of Hashes by :port and :protocol keys.
+  # Destination table looks like:
+  #   [
+  #     {
+  #      port: 80,
+  #      protocol: 'tcp',
+  #      extra: [ {source:, id: output:}, {source:, id: output:}]
+  #     },
+  #     ...
+  #   ]
+  # Both the destination hash and the new one have an :extra key, which is an
+  # array (of hashes with :source, :id and :output keys). If some of the hashes
+  # is not present in the destination row, it is added. If the destination row
+  # doesn't exist at all it is created.
+  #
+  # table   - the existing table where we want to merge a new entry.
+  #         May be nil, a Hash or an Array of hashes
+  # new_row - the new entry (a Hash) we want to put in the existing table.
+  #
+  # Returns table with new_row merged
+  def merge_service_extra(new_row)
+    table = self.properties[:services_extra]
```

Comment:
a few lines are shared with `merge_service` method. I think it's possible to DRY it up

Reply:
@aapomm tried to do so here: https://github.com/dradis/dradis-ce/pull/190/commits/1b028fcc49853720598f075c2e7d940fae35b797 
Suggestions?
I liked the idea to keep two separate methods, helps me understand what is going on.

Reply:
looks good 👍